src/cycles/eulerian.jl

```diff
@@ -0,0 +1,318 @@
+# Checking for existence of Eulerian circuit and Eulerian trail
+
+function single_nonzero_degree_component(
+    graph::AG,
+    ) where {T, AG <: AbstractGraph{T}}
+
+    visited = zeros(UInt8, nv(graph))
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            visited[v] = 2
+        end
+    end
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            continue
+        end
+        S = Vector{T}([v])
+        visited[v] = 1
+        while !isempty(S)
+            u = S[end]
+            w = 0
+            for n in outneighbors(graph, u)
+                if visited[n] == 0
+                    w = n
+                    break
+                end
+            end
+            if w != 0
+                visited[w] = 1
+                push!(S, w)
+            else
+                visited[u] = 2
+                pop!(S)
+            end
+        end
+        break
+    end
+    for v in vertices(graph)
+        visited[v] > 0 || return false
+    end
+    return true
+end
+
+"""
+    has_eulerian_trail(g)
+
+Return `true` if graph `g` contains [Eulerian trail](https://en.wikipedia.org/wiki/Eulerian_path).
+Note that a Eulerian Trail can have the same start and end vertex so all Eulerian Circuits are Trails as well.
+
+### Implementation Notes
+Uses single call to iterative version of DFS to check for the existence of a single component with non-zero degree in single_nonzero_degree_component().
+Condition for Eulerian trail :
+For undirected graph, if nodes with odd degree are at most 2.
+For directed graph,  if at most one vertex has (out-degree) − (in-degree) = 1, at most one vertex has (in-degree) − (out-degree) = 1.
+Since degree() accounts only once of self loop in directed graph, incremented degree by 1 for the algorithm to work in undirected graphs.
+"""
+function has_eulerian_trail end
+@traitfn function has_eulerian_trail(
+    graph::AG::(!IsDirected),
+    ) where {T, AG <: AbstractGraph{T}}
+
+    single_nonzero_degree_component(graph) || return false
+    odd = 0
+    for v in vertices(graph)
+        if (has_edge(graph, v, v) && iseven(outdegree(graph, v))) 
+            odd += 1
+        elseif (!has_edge(graph, v, v) && isodd(outdegree(graph, v)))
+            odd += 1
+        end
+    end
+
+    return (odd <= 2)
+end
+
+function has_eulerian_trail end
```

Comment:
You don't need to duplicate `function has_eulerian_trail end`, the reason for this is, to attach the docstring, as they don't play nicely with functions that are prefixed by `@traitfn`.